### DISCUSSION TITLE
Apicurio migration ReflectionAvroSerializer

### DISCUSSION THREAD

**Discussion by smeana** [Apicurio/apicurio-registry#2077] [General]
Hello 

I am testing up a possible migration from schema registry to Apicurio.
I cannot find the equivalent to `io.confluent.kafka.streams.serdes.avro.ReflectionAvroSerializer`
In case it is not implemented yet, is any straight forward way using Quarkus to serialize Avro without generating Avro friendly classes from an Avro schema?

Regards

**Comment by famarting** [comment]
Hi, we have that feature.

Check here https://www.apicur.io/registry/docs/apicurio-registry/2.1.x/getting-started/assembly-configuring-kafka-client-serdes.html#registry-serdes-types-avro_registry and look for `apicurio.registry.avro-datum-provider`

**Reply by smeana** [reply]
thanks @famartinrh 
Do you have any kafka client example using the `ReflectAvroDatumProvider`?
I cannot figure out how exactly should be added to the producer properties. 
I cannot find any property in `SerdeConfig` with the name of `apicurio.registry.avro-datum-provider`

tried:

```
configProps.putIfAbsent("apicurio.registry.avro-datum-provider",ReflectAvroDatumProvider.class);
configProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, AvroKafkaSerializer.class);
configProps.putIfAbsent(SerdeConfig.REGISTRY_URL, "http://10.8.0.1:8080");
```

**Reply by famarting** [reply]
we have this in apicurio-registry-examples repo https://github.com/Apicurio/apicurio-registry-examples/blob/master/avro-bean/src/main/java/io/apicurio/registry/examples/avro/bean/AvroBeanExample.java#L131

**Reply by smeana** [reply]
Many thanks @famartinrh 
Last question...
How we can integrate it with clients like kafka-tool, Kafka Ui tool-> https://github.com/provectus/kafka-ui
Reading the documentation Apicurio is fully compatible with schema registry but we can't manage to have it working just changing the schema registry URL for the Apicurio one (http://10.8.0.1:8085 -> http://10.8.0.1:8080/apis/registry/v2).
clients cannot deserialize but they cannot read the schemas too.

Regards

**Reply by carlesarnal** [reply]
@smeana the compatibility api lives under `/apis/ccompat/v6`.

**Reply by smeana** [reply]
Thanks @carlesarnal 
Using http://10.8.0.1:8080/apis/ccompat/v6 we get an Unknown magic byte issue.
Regards

**Reply by famarting** [reply]
thank you @smeana for using Apicurio Registry. I think your last issue about `Unknown magic byte issue` could be fixed by adding the property `apicurio.registry.as-confluent=true` to both your producers and consumers.

Apicurio uses 8 bytes to store the identifier in the kafka message body, while Confluent serdes use 4 bytes. We have the property `apicurio.registry.as-confluent` in our serdes to align that behavior.

**Reply by smeana** [reply]
Many thanks @famartinrh 
with the following config we still have the magic byte issue:

        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "10.8.0.1:9092");
        props.put(ProducerConfig.CLIENT_ID_CONFIG, "test-producer");
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, AvroKafkaSerializer.class.getName());
        props.putIfAbsent(ProducerConfig.ACKS_CONFIG, "all");
        props.putIfAbsent(SerdeConfig.REGISTRY_URL, "http://10.8.0.1:8080/apis/registry/v2");
        props.putIfAbsent(SerdeConfig.AUTO_REGISTER_ARTIFACT, Boolean.TRUE);
        props.putIfAbsent(SerdeConfig.ENABLE_CONFLUENT_ID_HANDLER, Boolean.TRUE);
        props.putIfAbsent(AvroKafkaSerdeConfig.AVRO_DATUM_PROVIDER, ReflectAvroDatumProvider.class.getName());

The Consumer is pointing to http://10.8.0.1:8080/apis/ccompat/v6 and using confluent avro deserializer

**Reply by famarting** [reply]
hmm that configuration should work I think... @smeana is it possible now you are getting that error from old messages when you started testing Apicurio? that could explain it... otherwise we would need to look more in depth into it...

**Reply by smeana** [reply]
Hello @famartinrh we tested it on new topics. I am trying to figure out what can be the problem.
regards

**Reply by EricWittmann** [reply]
Could you perhaps use e.g. kafkacat to validate that the magic byte is correct on the messages in the topic?  With the above config, the first byte of the message body should be 0x0 and then the next 4 bytes should be the schema's globalId.
